scenario_player/runner.py

```diff
@@ -153,279 +256,260 @@ def local_seed(self) -> str:
+            )
+            token_network_addresses.add(address)
 
-        :raises ScenarioError:
-            if ScenarioRunner.scenario.chain_name is not one of `('any', 'Any', 'ANY')`
-            and it is not a key in `chain_urls`.
-        """
-        chain_name = self.definition.settings.chain
-        if chain_name in ("any", "Any", "ANY"):
-            chain_name = random.choice(list(chain_urls.keys()))
+            log.info("Token Network Discovery", node=node._index, network=discovered)
 
-        log.info("Using chain", chain=chain_name)
-        try:
-            return chain_name, chain_urls[chain_name]
-        except KeyError:
-            raise ScenarioError(
-                f'The scenario requested chain "{chain_name}" for which no RPC-URL is known.'
+        if len(token_network_addresses) != 1:
+            raise RuntimeError(
+                f"Nodes diverged on the token network address, there should be "
+                f"exactly one token network available for all nodes. Current "
+                f"values : {token_network_addresses}"
             )
 
-    def wait_for_token_network_discovery(self, node) -> ChecksumAddress:
-        """Check for token network discovery with the given `node`.
+        return token_network_addresses.pop()
 
-        By default exit the wait if the token has not been discovered after `n` seconds,
-        where `n` is the value of :attr:`.timeout`.
+    def run_scenario(self):
+        self.node_controller.initialize_nodes()
+        node_starter = self.node_controller.start(wait=False)
+        node_monitor = self.node_controller.start_node_monitor()
+        node_addresses = self.node_controller.addresses
 
-        :raises TokenNetworkDiscoveryTimeout:
-            If we waited a set time for the token network to be discovered, but it wasn't.
-        """
-        log.info("Waiting till new network is found by nodes")
-        node_endpoint = API_URL_TOKEN_NETWORK_ADDRESS.format(
-            protocol=self.protocol, target_host=node, token_address=self.token.address
+        scenario = gevent.spawn(
+            self.setup_environment_and_run_main_task, node_addresses, node_starter
         )
 
-        started = time.monotonic()
-        elapsed = 0.0
-        while elapsed < self.definition.settings.timeout:
-            try:
-                resp = self.session.get(node_endpoint)
-                resp.raise_for_status()
+        # Wait for either a crash in one of the Raiden nodes or for the
+        # scenario to exit (successfully or not).
+        gevent.joinall({node_monitor, scenario}, raise_error=True, count=1)
 
-            except HTTPError as e:
-                # We explicitly handle 404 Not Found responses only - anything else is none
-                # of our business.
-                if e.response.status_code != 404:
-                    raise
+    def setup_environment_and_run_main_task(self, node_addresses, node_starter) -> None:
```

Comment:
This corresponds to the old [`ScenarioRunnner._setup`](https://github.com/raiden-network/scenario-player/blob/f208ed6cb4f713c1efc921b10458365bcf918b5d/scenario_player/runner.py#L237-L289). Renamed because in the end o f the setup function the scenario was executed.